### PR TITLE
Refactor

### PR DESCRIPTION
**Pull request by lee-woodridge**
Pulling out the logic from the Arena class into their own classes. Each part now has their own render function so they know what to do for themselves, which makes the render look succinct.

Also, I am Lee.

### CODE REVIEW

**Comment by lee-woodridge** [comment]
@cakey the diff seems to think that I added react etc although I merged your changes into master already :/

File: src/lib/player.coffee [hunk 1]
```diff
@@ -78,4 +82,119 @@ class Player
...
+        otherPs = _.reject _.values(@handler.players), team: @team
+
+        if Math.random() < Utils.game.speed(0.005) and not @startCastTime?
+            @handler.fire @, _.sample(otherPs).p, 'orb'
+
+        chanceToMove = Math.random() < Utils.game.speed(0.03)
+        if not @startCastTime? and (chanceToMove or @p.equal @destP)
+            @handler.moveTo @, @arena.map.randomPoint()
+
+    moveTo: (@destP) ->
```
Comment (cakey) [inline src/lib/player.coffee]: these functions fall through automatically, don't think you need to redefine then super

File: src/lib/player.coffee [hunk 2]
```diff
@@ -78,4 +82,119 @@ class Player
...
+        addEventListener "keypress", (event) =>
+
+            if skill = @keyBindings[String.fromCharCode event.which]
+                castP = @arena.map.mapMouseP.mapBound @p, @arena.map
+                @handler.fire @, castP, skill
+            else
+                console.log event
+                console.log event.which
+
+    update: (newTime) ->
```
Comment (cakey) [inline src/lib/player.coffee]: same here

File: src/lib/ui-element.coffee [hunk 3]
```diff
@@ -0,0 +1,18 @@
+class UIElement
```
Comment (cakey) [inline src/lib/ui-element.coffee]: What kind of logic might we add in the future? If we can;t think of anything then we probably don't need the base class yet.

Either way I would throw an error if it managed to full through?

File: src/client/app.coffee [hunk 4]
```diff
@@ -128,7 +44,7 @@ class Arena
             randomPoint = @map.randomPoint()
             randomTeam = _.sample((name for name, r of @teams))
 
-            @focusedUIPlayer = new UIPlayer @, @handler, randomPoint, randomTeam
+            @focusedUIPlayer = new UIPlayer this, @handler, randomPoint, randomTeam
```
Comment (cakey) [inline src/client/app.coffee]: why this?

Comment (lee-woodridge) [inline src/client/app.coffee]: I was using the style guide here https://github.com/polarmobile/coffeescript-style-guide#naming_conventions and it says:

Prefer @property over this.property.
return @property # Yes
return this.property # No

However, avoid the use of standalone @:
return this # Yes
return @ # No

I can change it back if you like?

File: src/lib/player.coffee (hunk 1 above)
Comment (lee-woodridge) [inline src/lib/player.coffee]: Oh really? I didn't know that, I'll test to make sure and then remove them?

File: src/lib/ui-element.coffee (hunk 3 above)
Comment (lee-woodridge) [inline src/lib/ui-element.coffee]: I'm not sure about the logic. I was thinking maybe we'd have a debug flag in the game config and could log drawing/updating etc in this base class.
The main draw for me was to be like "these are the functions a class should implement" so the error you suggest could help enforce this.

File: src/lib/player.coffee (hunk 1 above)
Comment (cakey) [inline src/lib/player.coffee]: yeah, take a look at the generated javascript if you're curious

File: src/client/app.coffee (hunk 4 above)
Comment (cakey) [inline src/client/app.coffee]: Fair enough, I prefer @. I think they prefer this as @method implies that @ includes the dot accessor... but to me @ just feels more coffeescripty

File: src/lib/ui-element.coffee (hunk 3 above)
Comment (cakey) [inline src/lib/ui-element.coffee]: you only call super in a few places though

**Pull request merged**